WIP: volume protection by pinning

tobwen opened the pull request:
#### Does this PR introduce a user-facing change?
Yes.

```release-note
* Added `--pinned` flag to `volume create` to create pinned volumes
* Added new `volume pin` commands to pin/unpin existing volumes
* Added `--include-pinned` flag to `volume rm` to allow removing pinned volumes
* Added pinned volume filtering support
* Added support for pinned volumes in `system prune` command
* Added HTTP API and ABI support for volume pinning
* Added runtime methods for handling pinned volumes and pruning
```

#### References
Reference https://github.com/containers/podman/issues/26807
Reference https://github.com/containers/podman/issues/23217

#### Actions required
- This is just the first step to demonstrate the function. Things like tunnel/remote implementation are still missing (it's just a stub).
- It needs to be discussed if we want to overload the CLI with `volume unpin` or rather use `volume pin --unpin` as supplied.

#### Note
Those are my first steps with the Podman code - please bear with me.

tobwen commented:
So sad... all the work for nothing.

mheon commented:
We generally don't review things until CI starts passing. In this case, looks like it's yelling about missing a manpage for the new commands? I can do a more thorough review early next week.

mheon commented on cmd/podman/volumes/pin.go:
```diff
@@ -0,0 +1,66 @@
+var (
+	pinDescription = `Mark or unmark a volume as pinned.
+
+Pinned volumes are excluded from system prune operations by default.`
```
System reset still catches them, yes?

tobwen replied:
That's a thing we should discuss... My main intention for pinning is to help inexperienced DevOps who are close to reaching capacity limits on their live systems, then execute `prune` or `reset` for some reason, and then realize that the volume containing their live database is gone and the backup is two days old.

After checking the implementation I found that Reset() was NOT respecting the pinned flag. Thanks, I fixed this.

tobwen replied:
I've added an optional solution for `reset`: https://github.com/containers/podman/pull/26846#issuecomment-3367584818

mheon commented on cmd/podman/system/prune.go:
```diff
@@ -137,7 +154,7 @@ func createPruneWarningMessage(pruneOpts entities.SystemPruneOptions) string {
 	return `WARNING! This command removes:
 	- all stopped containers
-	- all networks not used by at least one container%s%s
+	- all networks not used by at least one container%s%s` + pinnedNote + `
```
No need to conditionalize this IMO - just say "optionally excluding" instead of "excluding". We can trust folks to read the help text, I think.

tobwen replied:
Done. Removed the conditional logic and changed to a static "(excluding pinned volumes)" message that's always displayed.

mheon commented on libpod/runtime_volume.go:
```diff
@@ -141,3 +141,55 @@ func (r *Runtime) PruneVolumes(ctx context.Context, filterFuncs []VolumeFilter)
+
+// SetVolumePinned sets the pinned status of a volume by name.
+// Pinned volumes are excluded from system prune operations by default.
+func (r *Runtime) SetVolumePinned(volumeName string, pinned bool) error {
```
IMO, remove this - it doesn't add much over doing `runtime.GetVolume()` yourself and then calling `SetPinned()` on that

mheon commented on libpod/volume.go:
```diff
@@ -283,6 +286,38 @@ func (v *Volume) UsesVolumeDriver() bool {
 
+// Pinned returns whether this volume is marked as pinned.
+// Pinned volumes are excluded from system prune operations by default.
+func (v *Volume) Pinned() bool {
```
Rename to IsPinned()

mheon commented on libpod/volume.go:
```diff
@@ -283,6 +286,38 @@ func (v *Volume) UsesVolumeDriver() bool {
+		return err
+	}
+
+	v.state.Pinned = pinned
```
Should pinning a volume that is already pinned be an error?

tobwen replied:
Made it a no-op. If the volume is already in the desired pinned state, SetPinned() now returns early without error and without calling save(). This follows the principle of idempotency.

tobwen commented on pkg/domain/entities/types/volumes.go:
```diff
@@ -22,6 +22,9 @@ type VolumeCreateOptions struct {
 	GID *int `schema:"gid"`
+	// Pinned indicates that this volume should be excluded from
+	// system prune operations by default
+	Pinned bool `schema:"pinned"`
```
While `podman volume create` + `podman volume pin` would work, there's a window where the volume could be pruned between those two commands. I added a comment explaining this is to avoid race conditions when atomically creating and pinning a volume.

tobwen commented on libpod/options.go:
```diff
@@ -1753,6 +1753,20 @@ func withSetAnon() VolumeCreateOption {
 
+// WithVolumePinned sets the pinned flag for the volume.
+// Pinned volumes are excluded from system prune operations by default.
+func WithVolumePinned() VolumeCreateOption {
```
Done. Removed WithVolumePinned() and updated VolumeCreate to call SetPinned() on the volume after creation

tobwen commented:
### Additional feature

I've just added a new flag `--include-pinned` to `podman system reset` that allows users to optionally include pinned volumes in the reset operation.

Previously, pinned volumes were always excluded from being reset. This change:
- Updates warning messages to correctly reflect volume behavior
- Modifies the reset functionality to respect the new inclusion flag
- Propagates the new option through the system architecture
- Clarifies in the prune warning message that pinned volumes are optionally included

mheon commented:
@tobwen I think you're good to go ahead and write docs/tests for this - there's maintainer consensus that this is worth finishing and merging

tobwen commented:
> @tobwen I think you're good to go ahead and write docs/tests for this - there's maintainer consensus that this is worth finishing and merging

Nice! I already wrote [docs](https://github.com/containers/podman/pull/26846/commits/a0c6793153784304efc92ff5f5e9fe32d7b1377b), but [some checks don't like them](https://github.com/containers/podman/pull/26846/checks?check_run_id=53134804726).

I'll write some tests on the next days, but I think some actions like `Build Each Commit` will break checking.

tobwen committed:
feat: add pinned volume support to libpod core

Volumes can now be marked as "pinned" to protect them from
accidental removal during prune and reset operations.

Signed-off-by: tobwen <1864057+tobwen@users.noreply.github.com>

tobwen committed:
feat: wire pinned volume support through domain layer

Signed-off-by: tobwen <1864057+tobwen@users.noreply.github.com>

tobwen committed:
feat: expose pinned volumes through REST API and bindings

Signed-off-by: tobwen <1864057+tobwen@users.noreply.github.com>

tobwen committed:
feat: add volume pin/unpin commands and pinned flags to CLI

Signed-off-by: tobwen <1864057+tobwen@users.noreply.github.com>

tobwen committed:
test: add integration tests for pinned volumes

Signed-off-by: tobwen <1864057+tobwen@users.noreply.github.com>

tobwen committed:
docs: add documentation for pinned volumes

Signed-off-by: tobwen <1864057+tobwen@users.noreply.github.com>
